### DISCUSSION TITLE
Homogenisation System Issues

### DISCUSSION THREAD

**Discussion by Zoophish** [idaholab/moose#23483] [Q&A Modules: Solid mechanics] [answered]
Hi,

I'm trying to use the homogenisation system to extract the effective stiffnesses of a small hex-packed fibre composite RVE:

<img width="30%" alt="image" src="https://user-images.githubusercontent.com/3082590/220103373-e6539b81-8593-4bdd-821f-12a67920709e.png">

My issue is that the simulation gets stuck on `Setting Up Equation Systems....` before solving. I've given this process hours on a HPC to try and complete, but it never does.

The setup I am using is based off the input files in the Lagrangian/cartesian/homogenisation test folder. I've essentially just changed the materials and mesh file. I've tried using the master action and manual setup but both methods get stuck.

<details>
  <summary>My input file (manual setup):</summary>

  ```
mesh_file = 'micro.msh'
csv_file = 'cell_average_values'

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
  large_kinematics = false
  macro_gradient = hvar
  homogenization_constraint = homogenization
[]

[Variables]
  [disp_x]
  []
  [disp_y]
  []
  [disp_z]
  []
  [hvar]
    family = SCALAR
    order = NINTH
  []
[]

[UserObjects]
  [homogenization]
    type = HomogenizationConstraint
    constraint_types = 'stress strain strain strain stress strain strain strain strain'
    targets = 'stress11 strain21 strain31 strain12 stress22 strain32 strain13 strain23 strain33'
    execute_on = 'INITIAL LINEAR NONLINEAR'
  []
[]

[Kernels]
  [sdx]
    type = HomogenizedTotalLagrangianStressDivergence
    variable = disp_x
    component = 0
  []
  [sdy]
    type = HomogenizedTotalLagrangianStressDivergence
    variable = disp_y
    component = 1
  []
  [sdz]
    type = HomogenizedTotalLagrangianStressDivergence
    variable = disp_z
    component = 2
  []
[]

[ScalarKernels]
  [enforce]
    type = HomogenizationConstraintScalarKernel
    variable = hvar
  []
[]

[Mesh]
  [base]
    type = FileMeshGenerator
    file = ${mesh_file}
  []
[]

# [Modules]
#   [TensorMechanics]
#     [Master]
#       [all]
#         strain = SMALL
#         add_variables = true
#         new_system = true
#         formulation = TOTAL
#         volumetric_locking_correction = true
#         constraint_types = 'stress strain strain strain stress strain strain strain strain'
#         # targets = 'stress11 strain21 strain31 strain12 stress22 strain32 strain13 strain23 strain33'
#         targets = 'zero small small small zero small small small small'
#         generate_output = 'pk1_stress_xx pk1_stress_xy pk1_stress_xz pk1_stress_yx pk1_stress_yy pk1_stress_yz pk1_stress_zx pk1_stress_zy pk1_stress_zz deformation_gradient_xx deformation_gradient_xy deformation_gradient_xz deformation_gradient_yx deformation_gradient_yy deformation_gradient_yz deformation_gradient_zx deformation_gradient_zy deformation_gradient_zz cauchy_stress_xx cauchy_stress_xy cauchy_stress_xz cauchy_stress_yx cauchy_stress_yy cauchy_stress_yz cauchy_stress_zx cauchy_stress_zy cauchy_stress_zz strain_xx strain_xy strain_xz strain_yx strain_yy strain_yz strain_zx strain_zy strain_zz l2norm_pk1_stress'
#       []
#     []
#   []
# []

[Functions]
  [stress11]
    type = ParsedFunction
    value = '2e-12*t'
  []
  [stress22]
    type = ParsedFunction
    value = '1-12*t'
  []
  [strain33]
    type = ParsedFunction
    value = '8.0e-3*t'
  []
  [strain23]
    type = ParsedFunction
    value = '2.0e-3*t'
  []
  [strain13]
    type = ParsedFunction
    value = '-7.0e-3*t'
  []
  [strain12]
    type = ParsedFunction
    value = '1.0e-3*t'
  []
  [strain32]
    type = ParsedFunction
    value = '1.0e-3*t'
  []
  [strain31]
    type = ParsedFunction
    value = '2.0e-3*t'
  []
  [strain21]
    type = ParsedFunction
    value = '-1.5e-3*t'
  []
  [zero]
    type = ConstantFunction
    value = 0
  []
  [small]
    type = ConstantFunction
    value = 1e-9
  []
[]

[BCs]
  # periodic BCs
  # for complex periodic models, have to do
  # do this manually?
  [Periodic]
    # [x]
    #   variable = disp_x
    #   auto_direction = 'x y z'
    #   #   primary = 'front'
    # []
    # [y]
    #   variable = disp_y
    #   auto_direction = 'x y z'
    # []
    # [z]
    #   variable = disp_z
    #   auto_direction = 'x y z'
    # []
    [x]
      variable = disp_x
      primary = 'right'
      secondary = 'left'
      translation = '-22 0 0'
    []
    [y]
      variable = disp_y
      primary = 'top'
      secondary = 'bottom'
      translation = '0 -38.1051177665153 0'
    []
    [z]
      variable = disp_z
      primary = 'front'
      secondary = 'back'
      translation = '0 0 -38.1051177665153'
    []
  []

  # constraint BCs
  [fix_all_x]
    type = DirichletBC
    boundary = "fix_all"
    variable = disp_x
    value = 0
  []
  [fix_all_y]
    type = DirichletBC
    boundary = "fix_all"
    variable = disp_y
    value = 0
  []
  [fix_all_z]
    type = DirichletBC
    boundary = "fix_all"
    variable = disp_z
    value = 0
  []
  [fix_xy_x]
    type = DirichletBC
    boundary = "fix_xy"
    variable = disp_x
    value = 0
  []
  [fix_xy_y]
    type = DirichletBC
    boundary = "fix_xy"
    variable = disp_y
    value = 0
  []
  [fix_xy_z]
    type = DirichletBC
    boundary = "fix_xy"
    variable = disp_z
    value = 0
  []
  [fix_z]
    type = DirichletBC
    boundary = "fix_z"
    variable = disp_z
    value = 0
  []
[]

[Materials]
  [elastic_tensor_1]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 35e-3 # N/µm
    poissons_ratio = 0.389
    block = 'matrix'
  []
  [elastic_tensor_2]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 350e-3 # N/µm
    poissons_ratio = 0.21
    block = 'fibre'
  []
  [compute_stress]
    type = ComputeLagrangianLinearElasticStress
  []
  [compute_strain]
    type = ComputeLagrangianStrain
    homogenization_gradient_names = 'homogenization_gradient'
  []
  [compute_homogenization_gradient]
    type = ComputeHomogenizedLagrangianStrain
  []
[]

[Executioner]
  type = Transient
  solve_type = NEWTON
  line_search = none
  automatic_scaling = false
  compute_scaling_once = false
  l_max_its = 10000
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'
  l_tol = 1e-14
  nl_rel_tol = 1e-8
  nl_abs_tol = 5e-10

  start_time = 0.0
  dt = 0.1
  end_time = 1.0
[]

[Outputs]
  [mesh]
    type = Exodus
    file_base = ${csv_file}
  []
  [average_variable_values]
    type = CSV
    execute_on = 'TIMESTEP_END'
    file_base = ${csv_file}
  []
[]
  ```

</details>

Does this issue look familiar to anyone / any clues on what I am doing wrong here?

Thanks in advance,
Sam

**Comment by GiudGiud** [comment, marked as answer]
@hugary1995

**Reply by hugary1995** [reply]
This seems like a framework issue to me. @Zoophish: Does this behavior persist if you re-run the same simulation? We've had a similar issue on our cluster, but it was rather just random hangs -- could be at initializing equation system, a residual evaluation, or even an exodus output. We managed to fix the issue we encountered by using a different MPI.

Another possible cause of this issue is non-zero Jacobian allocations. Though I doubt that will make the equation system initialization system hang. IIRC we've already fixed the non-zero Jacobian allocation issue. But you could still try adding
```
[Problem]
  error_on_jacobian_nonzero_reallocation = true
[]
```
and see if it results in an error, which might give us more information.

The best way to figure out exactly what's going on is to diff the stack traces on all the processes. If one core has a unique stack trace, then that's most likely what leads to the hang. There is a script at https://github.com/idaholab/moose/blob/next/framework/scripts/find_hung_process.py that might help you get started.

**Reply by Zoophish** [reply]
Thanks for the quick reply.

We just updated to the next version of moose and it no longer seems to hang on the initialisation.

I am struggling to get it to converge in the non-linear iterations however. Do you have any advice on what could be causing this with the homogenisation system?

Thanks

**Reply by GiudGiud** [reply]
Some base advice
https://mooseframework.inl.gov/moose/application_usage/failed_solves.html
We would need to see the entire console log (attached if possible, not pasted) to say more

**Reply by hugary1995** [reply]
Hypre/amg doesn't work well on a saddle point problem in my experience. There might be a combination of options that can get hypre to work, but I've not found it yet. Field split should help in theory, but I also have had no luck with it yet.

However, a stronger preconditioner like lu or ilu have always worked well for me.

**Reply by hugary1995** [reply]
The input file you attached uses the default parallel lu package which is mumps I believe. That doesn't scale very well. Try asm and/or ilu.

**Reply by Zoophish** [reply]
Am I right in thinking that `lu` will be the best (given I can run it)? I tried `asm` and `ilu`, but they use up more iterations.

I also noticed that on the tilmestep before the solve fails, there is a stress concentration on the boundary:

<img width="50%" alt="image" src="https://user-images.githubusercontent.com/3082590/220390179-88159688-6356-4802-9514-2f07b6798f97.png">

This seems a bit bizarre to me - could it be an issue with the periodic BCs?

**Reply by hugary1995** [reply]
I think we need to see the colorbar range in the figure as well as the convergence log before making any further hypothesis.

**Reply by Zoophish** [reply]
Sure, here's the log.
[convergence_issues.out.zip](https://github.com/idaholab/moose/files/10795696/convergence_issues.out.zip)

**Reply by Zoophish** [reply]
And this is the strange stress concentration in X on the last time step it can solve:

<img width="60%" alt="image" src="https://user-images.githubusercontent.com/3082590/220577750-8024fcd2-3171-467b-b2cf-a91c390cd5df.png">

**Reply by hugary1995** [reply]
Trying to impose a constraint on the order of -12 is extremely challenging numerically. Please consider using a different unit system.

**Reply by Zoophish** [reply]
Okay that makes sense, I was trying out small values to see if it would affect the convergence.

I have removed the homogenisation constraints and it seems to be able to solve with zero strains and loads. Perhaps I am applying unphysical constraints? I would like to apply a small stress in yy, so to do this I am making the stress yy constraint $1 mN/μm^2$ and setting the rest to zero. Should this work in theory or have I misunderstood the constraint system?

Thanks,
Sam

**Reply by GiudGiud** [reply]
First I would turn off the line_search since it's failing
`line_search = none` in Executioner

then I would turn on automatic_scaling (also in Executioner block), because your residuals start very small and struggle to get smaller

**Reply by hugary1995** [reply]
> I would like to apply a small stress in yy, so to do this I am making the stress yy constraint and setting the rest to zero. Should this work in theory or have I misunderstood the constraint system?

Yes, that should work. Though you should be careful not to allow any rigid body modes, e.g., it happens when all your constraints are on stresses. Some of the constraints should act on the strain components.

I would start with @GiudGiud 's suggestions.

**Reply by Zoophish** [reply]
Thank you for your replies.

I'm mainly working off this input file example [here](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/test/tests/lagrangian/cartesian/total/homogenization/action/action_3d.i). This seems to be outdated as the ParsedFunctions are using `expression` rather than `value`. The target stresses also seem to be very specific here, though I cannot figure out exactly why these values were chosen.

I've made the changes to my executioner like GiudGiud suggested and still no convergence.

I am also pretty certain that the node constraints are correct (using fix_all, fix_xy and fix_z).

Another issue is that I can't seem to be able to use small strain as I get this error:
```*** ERROR ***
The following error occurred in the object "HomogenizationConstraints", of type "HomogenizationConstraintScalarKernel".

Homogenization kernel requires a variable of order 9 ```

**Reply by Zoophish** [reply]
Does it help to know I'm using TET4 elements? I've only found examples with this system of HEX8 elements.

**Reply by hugary1995** [reply]
The homogenization system should work with TET4 elements. If you see checkerboarding in the stress field, then there is an issue. But apparently there isn't.

The stress values in that input file were chosen arbitrarily for testing purposes.

For small strain, since the strain/stress tensors are symmetric, there will only be 6 independent constraints. There are examples for small strain as well.

At this point, if you can reproduce your problem with a much coarser mesh, I am happy to take a look at it. The easiest way for me to access your model is by sharing your branch with me.

**Reply by Zoophish** [reply]
The bit I was missing with small strain is to set the other entries to `none` instead of removing them. Maybe this could be documented for more clarity? I can run my model with small strain now on a much coarser mesh (linked below), but it can not converge on a more detailed mesh.

Hopefully this is coarse enough for you to run: 
[micro.msh.zip](https://github.com/idaholab/moose/files/10812441/micro.msh.zip)

This the input file I am using for finite strain:
[rve_action.i.zip](https://github.com/idaholab/moose/files/10812447/rve_action.i.zip)

The input file for small strain:
[rve_action 2.i.zip](https://github.com/idaholab/moose/files/10812811/rve_action.2.i.zip)

**Reply by hugary1995** [reply]
I downloaded your input file, and it is able to converge. Ideally I want a mesh that is as coarse as possible which can reproduce the convergence issue.

**Reply by hugary1995** [reply]
There were two issues in the large deformation input file. You should use `large_kinematics = true` with large deformation. I don't recommend using `compute_scaling_once = false` in any scenario.

<details>
  <summary>The modified input file (for large deformation):</summary>

```python
mesh_file = 'micro.msh'

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
  large_kinematics = true
[]

[Mesh]
  [base]
    type = FileMeshGenerator
    file = ${mesh_file}
  []
[]

[Modules]
  [TensorMechanics]
    [Master]
      [all]
        strain = FINITE
        add_variables = true
        new_system = true
        formulation = TOTAL
        volumetric_locking_correction = true
        constraint_types = 'strain stress stress stress stress stress stress stress stress'
        targets = 'small zero zero zero zero zero zero zero zero'
        generate_output = 'pk1_stress_xx pk1_stress_xy pk1_stress_xz pk1_stress_yx pk1_stress_yy pk1_stress_yz pk1_stress_zx pk1_stress_zy pk1_stress_zz deformation_gradient_xx deformation_gradient_xy deformation_gradient_xz deformation_gradient_yx deformation_gradient_yy deformation_gradient_yz deformation_gradient_zx deformation_gradient_zy deformation_gradient_zz cauchy_stress_xx cauchy_stress_xy cauchy_stress_xz cauchy_stress_yx cauchy_stress_yy cauchy_stress_yz cauchy_stress_zx cauchy_stress_zy cauchy_stress_zz strain_xx strain_xy strain_xz strain_yx strain_yy strain_yz strain_zx strain_zy strain_zz l2norm_pk1_stress'
      []
    []
  []
[]

[Functions]
  [zero]
    type = ConstantFunction
    value = 0
  []
  [small]
    type = ParsedFunction
    value = '1e-3*t'
  []
[]

[BCs]
  [Periodic]
    [x]
      variable = disp_x
      auto_direction = 'x y z'
    []
    [y]
      variable = disp_y
      auto_direction = 'x y z'
    []
    [z]
      variable = disp_z
      auto_direction = 'x y z'
    []
  []

  # constraint BCs
  [fix_all_x]
    type = DirichletBC
    boundary = "fix_all"
    variable = disp_x
    value = 0
  []
  [fix_all_y]
    type = DirichletBC
    boundary = "fix_all"
    variable = disp_y
    value = 0
  []
  [fix_all_z]
    type = DirichletBC
    boundary = "fix_all"
    variable = disp_z
    value = 0
  []
  [fix_xy_x]
    type = DirichletBC
    boundary = "fix_xy"
    variable = disp_x
    value = 0
  []
  [fix_xy_y]
    type = DirichletBC
    boundary = "fix_xy"
    variable = disp_y
    value = 0
  []
  [fix3_z]
    type = DirichletBC
    boundary = "fix_z"
    variable = disp_z
    value = 0
  []
[]

[Materials]
  [elastic_tensor_1]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 35e-3 # N/µm^2
    poissons_ratio = 0.389
    block = 'matrix'
  []
  [elastic_tensor_2]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 350e-3 # N/µm^2
    poissons_ratio = 0.21
    block = 'fibre'
  []
  [compute_stress]
    type = ComputeLagrangianLinearElasticStress
  []
[]

[Executioner]
  type = Transient
  solve_type = NEWTON
  line_search = none
  automatic_scaling = true
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'
  nl_rel_tol = 1e-8
  nl_abs_tol = 5e-10

  start_time = 0.0
  dt = 0.1
  end_time = 1.0
[]

[Outputs]
  exodus = true
[]

```
</details>

**Reply by Zoophish** [reply]
Thanks for your help, I'm going to mark this as solved at this point. I think I'll steer clear of running finite strain as a coarse mesh with small strain seems to work for now.
